README
VigMeUp for KDE
Version 1.0.3
http://www.c-nix.com/vigmeup

This program is free software.  Please read the COPYING file.
This software is copyright 1997, 1998, 1999, 2000 Sean R Melody and C-Nix Computing

The License can be found in the file COPYING
The Todo list can be found in the file TODO
The list of changes can be found in the file CHANGELOG
The instructions for installing this program can be found in the file INSTALL
The story behind the name can be found in the file STORY

Please direct all bugs reports and patches to vigmeup@c-nix.com as long as
development is current. Which it is as of June 19, 2000.  Thank you.

This is VigMeUp for KDE.  VigMeUp is an X11 program that uses the Qt widget set and the KDE
graphical user interface to create an alarm program that will play MPEG layer 3 audio (mp3s) files
at a specificed time.

Features:
1. Free Software
2. Plays mp3 files with your player (which must be installed for this version).
3. Saves last mp3 file played.
4. Saves last time of alarm.
5. Meets Open Source (TM) requirements.
6. Integerates into the K Desktop Environment.

VigMeUp for KDE is written in C++ using the Qt widget set and the KDE GUI.


ChangeLog
1.0.3 June 18, 2000
Well, stuff is working. I'm now using QLayouts to lay the geometry out
so it looks better when you resize it.  Also, some people were typing
in the name of the player.  It used to only set the player name if you
selected it from the file dialog.  Now it checks it immediately.


1.0.2 August 24, 1999
I did my best to fix the bug that was crippling this software.  I'm really
sorry about this...so now it works with x11amp, kmp3, mpg123, and most likely
others.  If you have another and get a positive report, let me know 
at vigmeup@c-nix.com
	
1.0.1 June 16, 1999
Quick bug fix.  I switched to a fork()/execv() execution of the mp3 player.
It appears to be a little cleaner than before.
Also, back to using KFileDialog, now that it actually works.
1.0.0 June 13, 1999
It's been a long time coming, but here it is, a version of VigMeUp
that is actually stable and KDE standards compliant.  Everything (that
I can think of) works.  The direction I'm taking it now is asthetic,
so don't worry about updating every second or whatever.  I used
KDevelop to bring it up to speed and make it KDE happy.
I changed the license BACK to GPL (From Artistic).
0.1.2 Mid-Winter... (Northern Hemisphere)
This was never released, but 0.1.2 had a bunch of changes that caused
VigMeUp to be very unstable.  I wasn't happy with it.
0.1.1 November 7
Added a spiffy menu bar and changed the snooze bar so
it is blue.  Changed the file dialog back to QFileDialog.  Developers
on the KDE-devel list were (once again) very unhelpful in even
bothering to find a solution to the problem. :(   The problem is that
KFileDialog will translate some special characters (like the
apostrophe) into its unicode equilvalent, like %29 or whatever.  A lot
of times, mp3s have weird names, so this can lead to problems.  If
anyone has any solutions, email me at vigmeup@lowrent.org.

0.1.0 September 3
Changed the CHANGELOG so the most recent are on the top.
Fixed a grammatical error in the CHANGELOG. (I had used errata to refer to one
error.)
Now I have an icon (from 0.07b) and a kdelnk file that will install when make
install runs.
The biggest changes are the new configuration options, ampProgram, and
snoozeLength.  Now YOU can select the program you use to play mp3s (Replay,
tkplay, x11amp, whatever!!) as long as it handles the first file argument as
the mp3.  There is now an on-board dialog to change the snooze length.  Both
of these options are saved with KConfig.
Changed numbering style similar to Linux kernel, this is development 0.1.0.
Use 0.07b if you are worried about stability, however, it still uses x11amp
and a snooze length of 10 minutes, but you can edit the source to change that.
Web pages are up at http://www.c-nix.com/vigmeup
Announced to freshmeat.net

0.07b August 05
Wonderful. Finally! Use KConfig and it works!
Last mp3 is saved and loaded on start up.
Also, KFileDialog is used instead of QFileDialog
This is the first version that is being released onto the net.

0.06b July 30
Read Forbes article on Free Software.
Fixes bug where /dev/dsp is locked by your amp program.
Fixed a scroll bar bug.

0.05b
Added a fixed width to the main dialog.

0.04b
Added mp3 Filter
Uses and needs KDE. This is a KDE Application, not a qt application.
Still depends on Qt though.

0.03b Bug Fixing, feature adding
Fixed space in file name errors.
Fixed apostrophe in file name errors.
Supports all file names now.

0.02b Secondary design
File selection enabled. Fill select any file and attempt to play it using x11amp. Therefore, a filter is needed.
Erratum: File names with an apostrophe (') will cause the system call to choke
and the alarm will not go off.

0.01b Initial Design Jun 11 1998. Night before two finals.
Alarm Clock like. Qt dependent. Doesn't need kde.






vigmeup/main.cpp
/***************************************************************************
                          main.cpp  -  description                              
                             -------------------                                         
    begin                : Sun Jun 13 03:59:48 CDT 1999
                                           
    copyright            : (C) 1999 by Sean Melody                         
    email                : vigmeup@c-nix.com                                     
 ***************************************************************************/

/***************************************************************************
 *                                                                         *
 *   This program is free software; you can redistribute it and/or modify  *
 *   it under the terms of the GNU General Public License as published by  *
 *   the Free Software Foundation; either version 2 of the License, or     *
 *   (at your option) any later version.                                   * 
 *                                                                         *
 ***************************************************************************/


#include "vigmeup.h" 
 
int main(int argc, char* argv[]) { 
  KApplication app(argc,argv,"vigmeup");  
 
  if (app.isRestored())
    { 
      RESTORE(VigmeupApp);
    }
  else 
    {
      VigmeupApp* vigmeup = new VigmeupApp;
      vigmeup->show();
    }  
  return app.exec();
}  
 


vigmeup/resource.h
/***************************************************************************
                          resource.h  -  description                              
                             -------------------                                         
    begin                : Sun Jun 13 03:59:48 CDT 1999
                                           
    copyright            : (C) 1999 by Sean Melody                         
    email                : vigmeup@c-nix.com                                     
 ***************************************************************************/

/***************************************************************************
 *                                                                         *
 *   This program is free software; you can redistribute it and/or modify  *
 *   it under the terms of the GNU General Public License as published by  *
 *   the Free Software Foundation; either version 2 of the License, or     *
 *   (at your option) any later version.                                   * 
 *                                                                         *
 ***************************************************************************/


#ifndef RESSOURCE_H
#define RESSOURCE_H

#ifdef HAVE_CONFIG_H
#include <config.h>
#endif

///////////////////////////////////////////////////////////////////
// resource.h  -- contains macros used for commands


///////////////////////////////////////////////////////////////////
// COMMAND VALUES FOR MENUBAR AND TOOLBAR ENTRIES


///////////////////////////////////////////////////////////////////
// File-menu entries
#define ID_FILE_CHOOSE_MP3                10030


#define ID_FILE_CHOOSE_PLAYER             10060
#define ID_FILE_QUIT											10070

///////////////////////////////////////////////////////////////////
// Edit-menu entries
#define ID_ACTION_SNOOZE                 11040


///////////////////////////////////////////////////////////////////
// View-menu entries                    
   #define ID_VIEW_STATUSBAR	12010
///////////////////////////////////////////////////////////////////
// Help-menu entries
#define ID_HELP                     1002

///////////////////////////////////////////////////////////////////
// General application values
#define ID_STATUS_MSG               1001

#define IDS_DEFAULT                 "Ready."

#define IDS_APP_ABOUT               "VigMeUp Version " VERSION "\nCopyright 1997, 1998, 1999\nSean R. Melody\n\nVisit VigMeUp on the Web at:\nhttp://www.c-nix.com/vigmeup\n\nEmail all comments, suggestions\nand virtual beers to\nvigmeup@c-nix.com"






///////////////////////////////////////////////////////////////////
// MACROS FOR THE CONNECT OF YOUR SIGNALS TO CORRESPONDENT SLOTS 
// IN YOUR MAIN IMPLEMENTATION OF MENUBAR AND TOOLBAR


///////////////////////////////////////////////////////////////////
// MENU CONNECTS
#define CONNECT_CMD(submenu)           connect(submenu, SIGNAL(activated(int)), SLOT(commandCallback(int)));connect(submenu,SIGNAL(highlighted(int)), SLOT(statusCallback(int)))


///////////////////////////////////////////////////////////////////
// TOOLBAR CONNECT  
#define CONNECT_TOOLBAR(ToolBar)        connect(ToolBar, SIGNAL(clicked(int)), SLOT(commandCallback(int)));connect(ToolBar, SIGNAL(highlighted(int,bool)), SLOT(statusCallback(int)))

///////////////////////////////////////////////////////////////////
// Create cases for entries and connect them with their functions
#define ON_CMD(id, cmd_class_function)   case id:cmd_class_function ;break;

///////////////////////////////////////////////////////////////////
// Create cases for entries and connect them to change statusBar entry
#define ON_STATUS_MSG(id, message)     case id:  slotStatusHelpMsg(message);break;

#endif // RESOURCE_H

















vigmeup/vigmeup.h
/***************************************************************************
                          vigmeup.h  -  description                              
                             -------------------                                         
    begin                : Sun Jun 13 03:59:48 CDT 1999
                                           
    copyright            : (C) 1999 by Sean Melody                         
    email                : vigmeup@c-nix.com                                     
 ***************************************************************************/

/***************************************************************************
 *                                                                         *
 *   This program is free software; you can redistribute it and/or modify  *
 *   it under the terms of the GNU General Public License as published by  *
 *   the Free Software Foundation; either version 2 of the License, or     *
 *   (at your option) any later version.                                   * 
 *                                                                         *
 ***************************************************************************/


#ifndef VIGMEUP_H
#define VIGMEUP_H 
 

#ifdef HAVE_CONFIG_H
#include <config.h>
#endif

//system includes
#include <sys/types.h>
#include <signal.h>    
#include <unistd.h>
#include <iostream.h>  
#include <errno.h>
// include files for QT
#include <qprinter.h>
#include <qpainter.h>

//old includes
#include <qdialog.h>
#include <qwidget.h>
#include <qlabel.h>
#include <qtimer.h>
#include <qlined.h>
#include <qscrbar.h>
#include <qpushbt.h>
#include <qtooltip.h>
#include <qlcdnum.h>
#include <kapp.h>
#include <qstring.h>
#include <qfiledlg.h>
#include <kmenubar.h>
#include <qpopmenu.h>
#include <qkeycode.h>
#include <kapp.h>

// include files for KDE
#include <kapp.h> 
#include <ktmainwindow.h> 
#include <kaccel.h>
#include <kiconloader.h>
#include <kmsgbox.h>

#include <kfiledialog.h>
// application specific includes

#include "resource.h"


// C system lib
#include <stdlib.h>

//Layout manager
#include <qlayout.h>
#include <kcontainer.h>
/**
  * This Class is the base class for your application. It sets up the main
  * window and reads the config file as well as providing a menubar, toolbar
  * and statusbar. For the main view, an instance of class VigmeupView is
  * created which creates your view.
  */
class VigmeupApp : public KTMainWindow 
{
  Q_OBJECT

    public:
  /** construtor */
  VigmeupApp(); 
  /** destructor */
  ~VigmeupApp();

  /** initMenuBar creates the menu_bar and inserts the menuitems */
  void initMenuBar();
  /** this creates the toolbars. Change the toobar look and add new toolbars in this
   * function */
  //  void initToolBar();
  /** setup the statusbar */
  void initStatusBar();
  /** setup the mainview*/
  //void initView();
  /** sets up the objects in the main window */
  void initObjects();
  /** save the app-specific options on slotAppExit or by an Options dialog */
  void saveOptions();
  /** read the app-specific options on init() or by an Options dialog */
  void readOptions();
  /** enable menuentries/toolbar items */
  void enableCommand(int id_);
  /** disable menuentries/toolbar items */
  void disableCommand(int id_);
  /** overloaded for Message box on last window exit */
  bool queryExit();

  //void initAlarmRunningFrame();
  bool alarmOff();
  void toggleAlarm ();
  void setAlarmRunning(bool b);
  bool alarmRunning;
  bool alarmOn();
  void playMp3(const char *mp3File, const char *mp3Player);
  QString mp3Player;
  QString lastPlayerDir;
  //QString mp3File;
  int snoozeLength;
  int childPID;
	
 
  public slots:
    /** switch argument for slot selection by menu or toolbar ID */
    void commandCallback(int id_);
  /** switch argument for Statusbar help entries on slot selection */
  void statusCallback(int id_);

  /** exits the application */
  void slotFileQuit();

  /** toggle the statusbar*/
  void slotViewStatusBar();
  /** change the status message to text */
  void slotStatusMsg(const char *text);
  /** change the status message of the whole statusbar temporary */
  void slotStatusHelpMsg(const char *text);

  void checkTime ();
  void chooseMp3();
  void snoozeIt();
  void choosePlayer();
  void setAlarm();
  void startAlarm();
  //QMenuBar *menuBar();
 private:


  /* main elements */
  QFrame *setAlarmFrame;

  KMenuBar *vmuMenuBar;
  KStatusBar *vmuStatusBar;

  QScrollBar *hourScrollBar;
  QScrollBar *minuteScrollBar;
  QScrollBar *snoozeScrollBar;
  QLCDNumber *hourLcd;
  QLCDNumber *minuteLcd;
  QLCDNumber *snoozeLcd;

  
  QLineEdit *playerTextBox;
  QPushButton *choosePlayerButton;
  QPushButton *chooseMp3Button;//button3;
  QPushButton *snoozeButton;//button1;
  QLabel *snoozeLengthLabel;//label3;

    QLineEdit *fileTextBox;//text1;
  QTimer *mainTimer;
  QLabel *colonLabel;

  /** file_menu contains all items of the menubar entry "File" */
  QPopupMenu *file_menu;
  /** action_menu contains all items of the menubar entry "Action" */
  QPopupMenu *action_menu;

  /** help_menu contains all items of the menubar entry "Help" */
  QPopupMenu *help_menu;
  /** flag if toolbar is visible or not. Used for kconfig and checking the
   * view-menu entry view toolbar. bViewStatusbar does the same for the
   * statusbar.
   */
  bool bViewToolbar;
  bool bViewStatusbar;

  /** used for KConfig to store and restore menubar position. Same with
   * tool_bar_pos. If more menubars or toolbars are created, you should add
   * positions as well and implement them in saveOptions() and readOptions().
   */
  KMenuBar::menuPosition menu_bar_pos;
  //KToolBar::BarPosition tool_bar_pos;
 protected:
 signals:
  //void changeHourLcd(int n);
  //void changeMinuteLcd2(int n);


};   
 
#endif // VIGMEUP_H
 
 
 
 
 
 
 
 























